### DISCUSSION TITLE
Read multiple capsules

### DISCUSSION THREAD

**Discussion by busslina** [GregoryConrad/rearch-rs#68] [General]
Can you show simple example of reading multiple capsules at once? I'm not using rust nigthly.

In the `fibonacci` example I see this: `*get.as_ref(Self(n - 1)) + get.as_ref(Self(n - 2))` which I don't understand why you dereference only one.

I'm getting this error reading two capsules: `cannot borrow "get" as mutable more than once at a time second mutable borrow occurs here`.

```rs
pub fn create_configuration_file_capsule(handle: CapsuleHandle) {
    let mut get = handle.get;
    let configuration_file = get.as_ref(configuration_file_capsule);

    // Already exists
    if configuration_file.exists() {
        panic!(
            "Configuration file already exists: {}",
            configuration_file.display()
        );
    }

    let do_create = Confirm::new()
        .with_prompt("Do you want to create the configuration file?")
        .interact()
        .unwrap();

    // drop(configuration_file);

    if !do_create {
        // drop(configuration_file);
        let mut console = get.as_ref(console_capsule).clone();

        console.write_fmt(format_args!(
            "Do not forget to supply the configuration file by yourself at '{}'\n",
            configuration_file.display()
        ));

        return;
    }

    todo!("Create configuration file");
}
```

**Comment by GregoryConrad** [comment]
> which I don't understand why you dereference only one

Probably to avoid the double `&mut` borrow; the `*` will copy (via `Copy`) the value that is behind the `&mut`.

As for reading multiple capsules; it's the same: copy or `Clone::clone` the value and then you can use `get` again.

**Reply by busslina** [reply]
Yes, I knew but I forget to clone it.

As for my question I got confused  because I read somewhere to "not read them individually". Now I see is on the `Container.read` docs. So seems this not apply to [`CapsuleReader`](https://docs.rs/rearch/latest/rearch/struct.Container.html#method.read):

```
Performs a consistent read on all supplied capsules that have cloneable data.

Consistency is important here: if you need the current data from a few different capsules,
do not read them individually, but rather group them together with one read() call. If you
read capsules one at a time, there will be increased overhead in addition to possible
inconsistency (say if you read one capsule and then the container is updated right after).
```

**Reply by busslina** [reply]
And also, shouldn't it worked by dropping the value just after using the handle for the second time (which seems it didn't work)?

**Reply by GregoryConrad** [reply]
That section of the docs only applies to container.read since CapsuleReader will always have a consistent snapshot of the container.

And yes, if you drop the first value read via get, then you should be able to call get again (after the drop).
